packages/api/db/functions.sql

```diff
@@ -0,0 +1,71 @@
+CREATE OR REPLACE FUNCTION create_perma_cache(data json) RETURNS TIMESTAMP
+    LANGUAGE plpgsql
+    volatile
+    PARALLEL UNSAFE
+AS
+$$
+DECLARE
+  inserted_at_ts TIMESTAMP;
+BEGIN
+    SET LOCAL statement_timeout = '30s';
+
+    insert into perma_cache (user_id, source_url, normalized_url, size)
+    values ((data ->> 'user_id')::BIGINT,
+            data ->> 'source_url',
+            data ->> 'normalized_url',
+            (data ->> 'size')::BIGINT)
+    returning inserted_at into inserted_at_ts;
+
+    insert into perma_cache_event (user_id, source_url, normalized_url, size, type, inserted_at)
+    values ((data ->> 'user_id')::BIGINT,
+            data ->> 'source_url',
+            data ->> 'normalized_url',
+            (data ->> 'size')::BIGINT,
+            ('Put')::perma_cache_event_type,
+            inserted_at_ts);
+
+    return inserted_at_ts;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION delete_perma_cache(data json) RETURNS TIMESTAMP
+    LANGUAGE plpgsql
+    volatile
+    PARALLEL UNSAFE
+AS
+$$
+DECLARE
+  deleted_at_ts TIMESTAMP;
+BEGIN
+    SET LOCAL statement_timeout = '30s';
+
+    WITH deleted AS (
+      delete from perma_cache
+      where user_id = (data ->> 'user_id')::BIGINT AND normalized_url = data ->> 'normalized_url'
+      returning source_url, size
```

Comment:
Just to be clear, we're only expecting 1 row here due to the `user_id,normalized_url UNIQUE` constraint?

Reply:
Yeah exactly